### PR TITLE
[Clang][Sema] Don't say "is declared here" for invalid template locations

### PR DESCRIPTION
**Pull request by bogner**
If a template is defined via an external AST source, it won't have a
location. When we emit warnings about misusing such templates we
shouldn't emit a "template is declared here" warning with no location,
as that's just confusing.

### CODE REVIEW

**Comment by bogner** [comment]
ping. @AaronBallman Does this look like a reasonable thing to do to you?

**Comment by bogner** [comment]
ping.

**Comment by cor3ntin** [comment]
That's a lot of repeated code. Should we have a "NoteTemplateLocation" function(s) of some sort that would do the check and the diag as appropriate?

**Comment by AaronBallman** [comment]
> ping. @AaronBallman Does this look like a reasonable thing to do to you?

Sorry for the delay on the review!

I agree that "is declared here" is useless when "here" is... nowhere. But the point to the note is to help the user to see what the actual declaration is to help them repair the issue. Losing the note entirely isn't the worst thing, but I wonder if it would be better for declarations coming from external AST sources to pretty print the declaration instead? e.g.,
Before:
```
hlsl::vector<1> BadVec; // expected-error{{template argument for template type parameter must be a type}}
// expected-note@*:* {{template is declared here}}
// expected-note@*:* {{template parameter is declared here}}
```
After:
```
hlsl::vector<1> BadVec; // expected-error{{template argument for template type parameter must be a type}}
// expected-note@-1 {{template declaration from hidden source: template <typename Ty> vector}}
// expected-note@-2 {{template parameter from hidden source: typename Ty}}
```
This would give the user more relevant information than just dropping the notes. (I have no idea if I'm asking for something incredibly difficult but I was thinking we could use the declaration printer to print this information out to the diagnostic in these cases.)

**Comment by bogner** [comment]
> That's a lot of repeated code. Should we have a "NoteTemplateLocation" function(s) of some sort that would do the check and the diag as appropriate?

Makes sense to me, and also makes @AaronBallman's idea of emitting notes referring to the external source a little easier. Done in the latest update.

**Comment by bogner** [comment]
> I agree that "is declared here" is useless when "here" is... nowhere. But the point to the note is to help the user to see what the actual declaration is to help them repair the issue. Losing the note entirely isn't the worst thing, but I wonder if it would be better for declarations coming from external AST sources to pretty print the declaration instead? e.g., Before:
> 
> ```
> hlsl::vector<1> BadVec; // expected-error{{template argument for template type parameter must be a type}}
> // expected-note@*:* {{template is declared here}}
> // expected-note@*:* {{template parameter is declared here}}
> ```
> 
> After:
> 
> ```
> hlsl::vector<1> BadVec; // expected-error{{template argument for template type parameter must be a type}}
> // expected-note@-1 {{template declaration from hidden source: template <typename Ty> vector}}
> // expected-note@-2 {{template parameter from hidden source: typename Ty}}
> ```
> 
> This would give the user more relevant information than just dropping the notes. (I have no idea if I'm asking for something incredibly difficult but I was thinking we could use the declaration printer to print this information out to the diagnostic in these cases.)

I made an attempt at this, and I think I like it. It does have a bit of novelty in terms of how it generates appropriate errors though.

File: clang/lib/Sema/SemaTemplate.cpp
Comment (cor3ntin) [inline clang/lib/Sema/SemaTemplate.cpp:7917]: Why not `<< Decl` directly?

**Review by AaronBallman — APPROVED** [review]
I like the behavior of this and the changes LGTM, but please give @ErichKeane a chance to weigh in as templates code owner before landing.

**Review by erichkeane — APPROVED** [review]
I think this is a very positive change, and see nothing besides Aaron's nit to comment on.  Thanks!

**Pull request merged**
